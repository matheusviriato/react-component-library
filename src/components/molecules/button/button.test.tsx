import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import { matchers } from 'jest-emotion'
import Button from '.'

expect.extend(matchers)

describe('Button component', () => {
  const id: string = 'robin-button-id'
  const mockFunction: jest.Mock<any, any> = jest.fn()

  describe('rendering tests', () => {
    test('should match snapshot', () => {
      const { baseElement } = render(
        <Button id={id} onClick={mockFunction}>
          Robin
        </Button>
      )
      expect(baseElement).toMatchSnapshot()
    })
    describe('when mount primary button', () => {
      test('should have a background color equal primary color by default', () => {
        render(
          <Button id={id} onClick={mockFunction}>
            Robin
          </Button>
        )
        expect(screen.getByTestId('button-testid')).toHaveStyleRule('background-color', '#ffcc00')
      })
    })
    describe('when mount outline button', () => {
      beforeEach(() => {
        render(
          <Button id={id} outline onClick={mockFunction}>
            Robin
          </Button>
        )
      })

      test('should have a background color equal white', () => {
        expect(screen.getByTestId('button-testid')).toHaveStyleRule('background-color', '#fff')
      })
      test('should have a border color equal darkGray', () => {
        expect(screen.getByTestId('button-testid')).toHaveStyleRule('border-color', '#333')
      })
    })
    describe('when mount disabled button', () => {
      beforeEach(() => {
        render(
          <Button id={id} disabled onClick={mockFunction}>
            Robin
          </Button>
        )
      })

      it('should have a background color equal lightestGray', () => {
        expect(screen.getByTestId('button-testid')).toHaveStyleRule('background-color', '#f5f5f5')
      })
      it('should change the cursor', () => {
        expect(screen.getByTestId('button-testid')).toHaveStyleRule('cursor', 'not-allowed')
      })
    })
  })
  describe('behavior tests', () => {
    it('should call the method passed as a callback when the button is clicked', () => {
      render(
        <Button id={id} onClick={mockFunction}>
          Robin
        </Button>
      )
      fireEvent.click(screen.getByTestId('button-testid'))
      expect(mockFunction).toHaveBeenCalled()
    })
    it('should not call the method passed as a callback when the button disabled is clicked', () => {
      const anotherMockFunction = jest.fn()
      render(
        <Button id={id} disabled onClick={anotherMockFunction}>
          Robin
        </Button>
      )
      fireEvent.click(screen.getByTestId('button-testid'))
      expect(anotherMockFunction).not.toHaveBeenCalled()
    })
  })
})
