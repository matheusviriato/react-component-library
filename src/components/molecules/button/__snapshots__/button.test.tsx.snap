// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component rendering tests should match snapshot 1`] = `
.emotion-0 {
  cursor: pointer;
  min-width: 160px;
  max-height: 48px;
  font-size: 16px;
  color: #333;
  box-shadow: #00000029 0px 2px 2px 0px;
  border-radius: 24px;
  border: none;
  padding: 15px;
  font-weight: 600;
  text-transform: uppercase;
  background-color: #ffcc00;
}

.emotion-0:hover {
  box-shadow: #00000033 0 3px 8px 0;
}

.emotion-0:disabled {
  border: 1px solid #ccc;
  background-color: #f5f5f5;
  color: #ccc;
  cursor: not-allowed;
}

.emotion-0:disabled:hover {
  box-shadow: #00000029 0px 2px 2px 0px;
}

<body>
  <div>
    <button
      class="emotion-0"
      data-testid="button-testid"
      id="robin-button-id"
      type="submit"
    >
      Robin
    </button>
  </div>
</body>
`;
